Fix benchmark script to work with poetry

bachase opened the pull request:
When initially refactoring to use poetry, the Dockerfile used in the benchmark task stored the virtual env directory within the package working directory.

However, the github actions benchmark script maps the runner's working copy of the ucc directory to the dockerfile run of the benchmark script. This shadows the /ucc/.venv directory.

This fix switches to store the virtual environment directory outside the package working directory to avoid this shadowing.

Fixes #212

bachase commented:
Note that I manually ran the benchmark actions on this commit to ensure it worked (you can see that in https://github.com/unitaryfund/ucc/actions/runs/13166142638) and the commit it created e7efdfa.

After review and before merging, I may force rebase to scrub that commit and combine the gitignore fix changeset with the other.

bachase commented on benchmarks/scripts/__pycache__/common.cpython-312.pyc:
I noticed this file somehow got checked in even with .gitgnore and seems to have been around for a while. As I learn more on how the benchmark code works, can see how to ensure this is filtered.

Misty-W commented:
I tried testing locally per instructions in `.github/workflows/ucc-benchmarks.yml`, but I get an error message 

> Error: Could not find any stages to run. View the valid jobs with `act --list`. Use `act --help` to find how to filter by Job ID/Workflow/Event Name

Trying `act push --container-architecture linux/amd64 -j run-benchmarks` (as implied by the output of `act --list` gives a warning message before abruptly exiting:

> Skipping unsupported platform -- Try running with `-P ucc-benchmarks-8-core-U22.04=...`

@natestemen do you get the same errors?

bachase commented:
You are correct @Misty-W , I forgot I had to make 2 other edits to get it to run locally. 

1. I had to change `runs-on: ucc-benchmarks-8-core-U22.04` to `runs-on: ubuntu-latest` since that specific ucc-benchmarks runner is not a standard kind that act knows about.
2. In the docker run step, I had to change ` -v "/home/runner/work/ucc/ucc:/ucc" \` to ` -v "{path to ucc repo on my laptop}:/ucc" \`

I meant to try and add some filters to the github actions like the below but it was proving tricky https://github.com/unitaryfund/ucc/blob/0e7ddafd178db376f48ec8b8afe43ada55d8876e/.github/workflows/ucc-benchmarks.yml#L50

Misty-W commented:
Thanks @bachase, I  made the edits in your [comment above](https://github.com/unitaryfund/ucc/pull/214#issuecomment-2640863762), but encountered a new error:

>  => ERROR [stage-0 6/8] RUN --mount=type=cache,target=/tmp/poetry_cache   10.2s
PEP517 build of a dependency failed
Note: This error originates from the build backend, and is likely not a problem with poetry but one of the following issues with pyqasm (0.1.0)
      - not supporting PEP 517 builds
      - not specifying PEP 517 build requirements correctly
      -  the build requirements are incompatible with your operating system or Python version
      - the build requirements are missing system dependencies (eg: compilers, libraries, headers).

bachase commented:
@Misty-W I couldn't get it working locally for myself either, but timeboxed to 30 minutes. 

Instead, I removed those instructions for running locally and reworked some of the githistory. I've rerun the GH action on this branch manually and you can see the successful commit added 06a00fe.

Assuming this now looks good, I'll scrub that commit before merging, and then create a new issue for reviewing the benchmarking code for portability.

Misty-W reviewed (approved):
Given the successful benchmark run, this is ready to merge once the benchmark results commit is scrubbed. thanks for the quick resolution @bachase!
